INI Plugin: Tests Fail on OS X

Since pull request #599 was merged `testscr_check_get_set` and `testmod_ini` fail on my machine. Here is the output of `testmod_ini`:

```
INI	   TESTS
==================

Compare <[section1]
>, with <section1/key1 = value1
>
in file /usr/local/share/elektra/test_data/ini/plainini, line 2.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:78: error in test_plainIniWrite: files do not match as expected
Compare <[section1]
>, with <section1/key1 = value1
>
in file /usr/local/share/elektra/test_data/ini/emptyval, line 2.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:106: error in test_plainIniEmptyWrite: files do not match as expected
Compare <key1 = value1
>, with <[section1]
>
in file /usr/local/share/elektra/test_data/ini/sectionini, line 5.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:347: error in test_sectionWrite: files do not match as expected
Compare <[section1]
>, with <section1/key1 = val1
>
in file /usr/local/share/elektra/test_data/ini/sectionmerge.output, line 1.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:386: error in test_sectionMerge: files do not match as expected
Compare <[Section1]
>, with <Section1/testkey0 = testval0
>
in file /usr/local/share/elektra/test_data/ini/insertTest.output.ini, line 1.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:444: error in test_insertOrder: files do not match as expected
Compare <array = 0
>, with <array = 2
>
in file /usr/local/share/elektra/test_data/ini/arrayInsertOut.ini, line 4.
tests/cframework/tests.c:174: error in compare_line_files: comparing lines failed
src/plugins/ini/testmod_ini.c:492: error in test_arrayInsert: files do not match as expected

test_ini RESULTS: 191 test(s) done. 12 error(s).
```

The command `ctest -V -R testscr_check_get_set` prints the following output:

```
UpdateCTestConfiguration  from :build/DartConfiguration.tcl
Parse Config file:build/DartConfiguration.tcl
UpdateCTestConfiguration  from :build/DartConfiguration.tcl
Parse Config file:build/DartConfiguration.tcl
Test project build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 77
    Start 77: testscr_check_get_set

77: Test command: /Users/rene/Dropbox/Studium/Master\ Thesis/Configuration\ Parsing/Source/Elektra/build/tests/shell/testscr_check_get_set.sh
77: Test timeout computed to be: 1500
77:
77: ELEKTRA BASIC COMMAND SCRIPTS TESTS
77:
77: Check if script tests the correct version
77: augeas not a read-write storage
77: ccode not a read-write storage
77: conditionals not a read-write storage
77: constants not a read-write storage
77: counter not a read-write storage
77: crypto_gcrypt not a read-write storage
77: csvstorage not a read-write storage
77: curlget not a read-write storage
77: doc not a read-write storage
77: dpkg not a read-write storage
77:
77: -- Testing dump
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: enum not a read-write storage
77: error not a read-write storage
77: filecheck not a read-write storage
77: glob not a read-write storage
77: hexcode not a read-write storage
77: hidden not a read-write storage
77: hosts not a read-write storage
77: iconv not a read-write storage
77:
77: -- Testing ini
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: Did not find keyerror: user/tests/script/hello/a/array/#3 is not 3
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58334 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#4 is not 4
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58337 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#5 is not 5
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58339 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#6 is not 6
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58341 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#7 is not 7
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58343 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#8 is not 8
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58348 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key user/tests/script/hello/a/array/#0
77: Did not find keyerror: user/tests/script/hello/a/array/#9 is not 9
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: Did not find keyerror: system/tests/script/hello/a/array/#3 is not 3
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58388 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#4 is not 4
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58390 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#5 is not 5
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58392 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#6 is not 6
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58394 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#7 is not 7
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58399 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#8 is not 8
77: build/tests/shell/testscr_check_get_set.sh: line 1134: 58401 Segmentation fault: 11  "$KDB" set "$KEY" "$VALUE" > /dev/null
77: error: could not set key system/tests/script/hello/a/array/#0
77: Did not find keyerror: system/tests/script/hello/a/array/#9 is not 9
77: error: found remaining files /Users/rene/.config/test.ini.58334:1461763871.4295607571.tmp /Users/rene/.config/test.ini.58337:1461763872.4295090567.tmp /Users/rene/.config/test.ini.58339:1461763872.4295528148.tmp /Users/rene/.config/test.ini.58341:1461763873.4294996850.tmp /Users/rene/.config/test.ini.58343:1461763873.4295381935.tmp /Users/rene/.config/test.ini.58348:1461763873.4295812303.tmp in /Users/rene/.config
77: error: found remaining files /etc/kdb/test.ini.58388:1461763874.140733194299509.tmp /etc/kdb/test.ini.58390:1461763875.140733193704383.tmp /etc/kdb/test.ini.58392:1461763875.140733194182024.tmp /etc/kdb/test.ini.58394:1461763876.140733193547216.tmp /etc/kdb/test.ini.58399:1461763876.140733193990036.tmp /etc/kdb/test.ini.58401:1461763877.140733193397306.tmp in /etc/kdb
77: iterate not a read-write storage
77: jni not a read-write storage
77: keytometa not a read-write storage
77: line not a read-write storage
77: lineendings not a read-write storage
77: list not a read-write storage
77: logchange not a read-write storage
77: lua not a read-write storage
77: mathcheck not a read-write storage
77: network not a read-write storage
77:
77: -- Testing ni
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: noresolver not a read-write storage
77: null not a read-write storage
77: path not a read-write storage
77: profile not a read-write storage
77: python2 not a read-write storage
77: regexstore not a read-write storage
77: rename not a read-write storage
77: resolver_fm_hpu_b not a read-write storage
77: resolver_fm_b_b not a read-write storage
77: resolver_fm_pb_b not a read-write storage
77: resolver_fm_hb_b not a read-write storage
77: resolver_fm_hp_b not a read-write storage
77: resolver_fm_ub_x not a read-write storage
77: resolver_fm_xb_x not a read-write storage
77: resolver_fm_xp_x not a read-write storage
77: resolver_fm_xhp_x not a read-write storage
77: resolver_fm_uhb_xb not a read-write storage
77: semlock not a read-write storage
77: shell not a read-write storage
77: spec not a read-write storage
77: struct not a read-write storage
77: sync not a read-write storage
77: syslog not a read-write storage
77:
77: -- Testing tcl
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: template not a read-write storage
77: timeofday not a read-write storage
77: tracer not a read-write storage
77: type not a read-write storage
77: uname not a read-write storage
77: validation not a read-write storage
77: wresolver not a read-write storage
77:
77: -- Testing xmltool
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77:
77: -- Testing yajl
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: testing ls command
77: testing rm command
77: Did not find keyDid not find keytesting array
77: -n Master
77: Configuration
77: testscr_check_get_set.sh RESULTS: 538 test(s) done
77:  28 error(s).
1/1 Test #77: testscr_check_get_set ............***Failed   17.89 sec

0% tests passed, 1 tests failed out of 1

Label Time Summary:
kdbtests    =  17.89 sec (1 test)
memleak     =  17.89 sec (1 test)

Total Test time (real) =  17.90 sec

The following tests FAILED:
	 77 - testscr_check_get_set (Failed)
Errors while running CTest
```

If you need any additional information, then please just comment below.

--- a/src/plugins/ini/ini.c
+++ b/src/plugins/ini/ini.c
@@ -209,12 +209,12 @@ static void iniBomHandler (void * vhandle, short BOM)
 
 static void setKeyOrderNumber (Key * sectionKey, Key * key)
 {
-	const Key * childMeta = keyGetMeta (sectionKey, "lastChild");
-	keySetMeta (key, "keyNo", keyString (childMeta));
+	const Key * childMeta = keyGetMeta (sectionKey, "ini/key/last");
+	keySetMeta (key, "ini/key/number", keyString (childMeta));
 	Key * newChild = keyDup (childMeta);
 	keyAddName (newChild, keyString (newChild));
 	elektraArrayIncName (newChild);
-	keySetMeta (sectionKey, "lastChild", keyBaseName (newChild));
+	keySetMeta (sectionKey, "ini/key/last", keyBaseName (newChild));
 	keyDel (newChild);
 	keySetMeta (key, "order", keyString (keyGetMeta (sectionKey, "order")));
 }
@@ -285,7 +285,7 @@ static void insertKeyIntoKeySet (Key * parentKey, Key * key, KeySet * ks)
 	keySetMeta (key, "parent", parent);
 	if (keyIsBinary (key))
 	{
-		keySetMeta (key, "lastChild", "#0");
+		keySetMeta (key, "ini/key/last", "#0");
 		Key * cutKey = keyNew (parent, KEY_END);
 		KeySet * cutKS = ksCut (ks, cutKey);
 		Key * cur;
@@ -364,11 +364,11 @@ static int iniKeyToElektraKey (void * vhandle, const char * section, const char 
 		if (!sectionKey)
 		{
 			keySetMeta (appendKey, "order", "#0");
-			keySetMeta (appendKey, "lastChild", "#0");
+			keySetMeta (appendKey, "ini/key/last", "#0");
 			keySetBinary (appendKey, 0, 0);
 			ksAppendKey (handle->result, keyDup (appendKey));
 			keySetMeta (appendKey, "order", "#0");
-			keySetMeta (appendKey, "lastChild", "#0");
+			keySetMeta (appendKey, "ini/key/last", "#0");
 			keySetMeta (appendKey, "binary", 0);
 			sectionKey = ksLookup (handle->result, appendKey, KDB_O_NONE);
 		}
@@ -476,7 +476,7 @@ static int iniSectionToElektraKey (void * vhandle, const char * section)
 		return 1;
 	}
 	setOrderNumber (handle->parentKey, appendKey);
-	keySetMeta (appendKey, "lastChild", "#0");
+	keySetMeta (appendKey, "ini/key/last", "#0");
 	keySetBinary (appendKey, 0, 0);
 	flushCollectedComment (handle, appendKey);
 	ksAppendKey (handle->result, appendKey);
@@ -872,7 +872,7 @@ void createGlobalRoot (Key * parentKey, KeySet * ks)
 	keySetMeta (appendKey, "binary", 0);
 	keyAddName (appendKey, INTERNAL_ROOT_SECTION);
 	keySetMeta (appendKey, "order", "#0");
-	keySetMeta (appendKey, "lastChild", "#0");
+	keySetMeta (appendKey, "ini/key/last", "#0");
 	ksAppendKey (ks, appendKey);
 }
 
@@ -890,7 +890,7 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 			keySetMeta (arrayParent, "ini/array", keyBaseName (newKey));
 			keySetMeta (newKey, "ini/arrayMember", "");
 			keySetMeta (newKey, "order", keyString (keyGetMeta (arrayParent, "order")));
-			keySetMeta (newKey, "keyNo", 0); // keyBaseName (newKey));
+			keySetMeta (newKey, "ini/key/number", 0); // keyBaseName (newKey));
 			ksAppendKey (newKS, newKey);
 		}
 	}
@@ -898,7 +898,7 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 	{
 		const char * oldVal = keyString (arrayParent);
 		keySetMeta (arrayParent, "ini/array", keyBaseName (cur));
-		keySetMeta (arrayParent, "keyNo", 0);
+		keySetMeta (arrayParent, "ini/key/number", 0);
 		if (oldVal && strlen (oldVal))
 		{
 			Key * arrayInitKey = keyDup (arrayParent);
@@ -908,11 +908,11 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 			keySetMeta (arrayInitKey, "ini/arrayMember", "");
 			ksAppendKey (newKS, arrayInitKey);
 			keySetMeta (arrayInitKey, "order", keyString (keyGetMeta (arrayParent, "order")));
-			keySetMeta (arrayInitKey, "keyNo", 0); // keyBaseName(arrayInitKey));
+			keySetMeta (arrayInitKey, "ini/key/number", 0); // keyBaseName(arrayInitKey));
 		}
 		ksAppendKey (newKS, newKey);
 		keySetMeta (newKey, "order", keyString (keyGetMeta (arrayParent, "order")));
-		keySetMeta (newKey, "keyNo", 0); // keyBaseName(newKey));
+		keySetMeta (newKey, "ini/key/number", 0); // keyBaseName(newKey));
 		keySetMeta (newKey, "ini/arrayMember", "");
 	}
 	else if (arrayParent && (keyBaseName (newKey)[1] == '0'))
@@ -931,8 +931,8 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 			keySetMeta (arrayParent, "ini/array", keyBaseName (cur));
 			ksAppendKey (newKS, arrayParent);
 			insertKeyIntoKeySet (parentKey, arrayParent, newKS);
-			keySetMeta (arrayParent, "lastChild", 0);
-			keySetMeta (arrayParent, "keyNo", 0);
+			keySetMeta (arrayParent, "ini/key/last", 0);
+			keySetMeta (arrayParent, "ini/key/number", 0);
 		}
 		keySetBinary (newKey, 0, 0);
 		keySetMeta (arrayParent, "ini/array", keyBaseName (cur));
@@ -941,8 +941,8 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 		insertKeyIntoKeySet (parentKey, newKey, newKS);
 		keySetMeta (newKey, "binary", 0);
 		keySetString (newKey, keyString (cur));
-		keySetMeta (newKey, "lastChild", 0);
-		keySetMeta (newKey, "keyNo", 0);
+		keySetMeta (newKey, "ini/key/last", 0);
+		keySetMeta (newKey, "ini/key/number", 0);
 	}
 	else if (keyIsDirectBelow (parentKey, newKey))
 	{
@@ -1037,8 +1037,8 @@ static int iniCmpOrder (const void * a, const void * b)
 
 	const Key * kaom = keyGetMeta (ka, "order");
 	const Key * kbom = keyGetMeta (kb, "order");
-	const Key * kakm = keyGetMeta (ka, "keyNo");
-	const Key * kbkm = keyGetMeta (kb, "keyNo");
+	const Key * kakm = keyGetMeta (ka, "ini/key/number");
+	const Key * kbkm = keyGetMeta (kb, "ini/key/number");
 
 	int ret = keyGetNamespace (ka) - keyGetNamespace (kb);
 	if (!ret)

--- a/src/plugins/ini/ini.c
+++ b/src/plugins/ini/ini.c
@@ -1042,9 +1042,18 @@ static int iniCmpOrder (const void * a, const void * b)
 	const Key * kbkm = keyGetMeta (kb, "keyNo");
 
 	int ret = keyGetNamespace (ka) - keyGetNamespace (kb);
-	if (!ret) ret = strcmp (keyString (kaom), keyString (kbom));
 	if (!ret)
 	{
+		if (!kaom && !kbom) return 0;
+		if (kaom && !kbom) return 1;
+		if (!kaom && kbom) return -1;
+		ret = strcmp (keyString (kaom), keyString (kbom));
+	}
+	if (!ret)
+	{
+		if (!kakm && kbkm) return -1;
+		if (!kakm && !kbkm) return strcmp (keyName (ka), keyName (kb));
+		if (kakm && !kbkm) return 1;
 		if (kakm && kbkm) ret = strcmp (keyString (kakm), keyString (kbkm));
 	}
 

--- a/src/plugins/ini/ini.c
+++ b/src/plugins/ini/ini.c
@@ -237,7 +237,6 @@ static int iniKeyToElektraArray (CallbackHandle * handle, Key * existingKey, Key
 		keySetString (appendKey, value);
 		keySetMeta (appendKey, "ini/arrayMember", "");
 		keySetMeta (appendKey, "order", keyString (keyGetMeta (existingKey, "order")));
-		keySetMeta (appendKey, "keyNo", keyBaseName (appendKey));
 		ksAppendKey (handle->result, appendKey);
 		keySetMeta (existingKey, "ini/array", keyBaseName (appendKey));
 		ksAppendKey (handle->result, existingKey);
@@ -263,7 +262,6 @@ static int iniKeyToElektraArray (CallbackHandle * handle, Key * existingKey, Key
 			return -1;
 		}
 		keySetString (appendKey, origVal);
-		keySetMeta (appendKey, "keyNo", keyBaseName (appendKey));
 		ksAppendKey (handle->result, keyDup (appendKey));
 		free (origVal);
 		if (elektraArrayIncName (appendKey) == -1)
@@ -272,7 +270,6 @@ static int iniKeyToElektraArray (CallbackHandle * handle, Key * existingKey, Key
 		}
 		keySetMeta (appendKey, "parent", 0);
 		keySetString (appendKey, value);
-		keySetMeta (appendKey, "keyNo", keyBaseName (appendKey));
 		ksAppendKey (handle->result, keyDup (appendKey));
 		keyDel (appendKey);
 		keyDel (sectionKey);
@@ -901,7 +898,7 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 	{
 		const char * oldVal = keyString (arrayParent);
 		keySetMeta (arrayParent, "ini/array", keyBaseName (cur));
-
+		keySetMeta (arrayParent, "keyNo", 0);
 		if (oldVal && strlen (oldVal))
 		{
 			Key * arrayInitKey = keyDup (arrayParent);
@@ -935,6 +932,7 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 			ksAppendKey (newKS, arrayParent);
 			insertKeyIntoKeySet (parentKey, arrayParent, newKS);
 			keySetMeta (arrayParent, "lastChild", 0);
+			keySetMeta (arrayParent, "keyNo", 0);
 		}
 		keySetBinary (newKey, 0, 0);
 		keySetMeta (arrayParent, "ini/array", keyBaseName (cur));
@@ -944,6 +942,7 @@ void arrayHandler (Key * parentKey, Key * newKey, Key * cur, Key * sectionKey, K
 		keySetMeta (newKey, "binary", 0);
 		keySetString (newKey, keyString (cur));
 		keySetMeta (newKey, "lastChild", 0);
+		keySetMeta (newKey, "keyNo", 0);
 	}
 	else if (keyIsDirectBelow (parentKey, newKey))
 	{

